{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc656fbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import multiprocessing\n",
    "import operator\n",
    "import os\n",
    "import pickle\n",
    "from itertools import product\n",
    "\n",
    "import joblib\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotx\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pyarrow.feather as feather\n",
    "import scipy\n",
    "import seaborn as sns\n",
    "import shap\n",
    "import xgboost as xgb\n",
    "from scipy.stats import gaussian_kde\n",
    "from adjustText import adjust_text\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.datasets import load_svmlight_file\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "from sklearn.inspection import partial_dependence\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from sklearn.metrics import root_mean_squared_error, silhouette_score\n",
    "from sklearn.model_selection import (\n",
    "    GridSearchCV,\n",
    "    KFold,\n",
    "    cross_val_score,\n",
    "    cross_validate,\n",
    "    train_test_split,\n",
    ")\n",
    "from sklearn.pipeline import Pipeline\n",
    "from kneed import KneeLocator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a03d3d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "plt.rcParams['font.family'] = 'Arial'\n",
    "gene_select_dir = \"/dcs05/hongkai/data/next_cutntag/bulk/dna_methylation/cpg_island\"\n",
    "gene_select_filename = \"gene_categories.json\"\n",
    "gene_select_dir_filename = os.path.join(gene_select_dir, gene_select_filename)\n",
    "with open(gene_select_dir_filename) as fp:\n",
    "    gene_select_dict = json.load(fp)\n",
    "gene_select_dict[\"coding_all\"] = gene_select_dict[\"coding_cpg\"] + gene_select_dict[\"coding_non_cpg\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "932f46fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "target_pair_mapping_file_36x = \"/dcs05/hongkai/data/next_cutntag/script/utils/target_pair_short_hand.csv\"\n",
    "target_pair_mapping_df = pd.read_csv(target_pair_mapping_file_36x, sep=\" \")\n",
    "target_pairs_selected_file = \"/dcs05/hongkai/data/next_cutntag/bulk/explainability/pdp_explain_results/target_pairs_selected.json\"\n",
    "\n",
    "with open(target_pairs_selected_file) as fp:\n",
    "    target_pairs = json.load(fp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccf46a81",
   "metadata": {},
   "outputs": [],
   "source": [
    "# helper functions\n",
    "def map_target_names(target_pair_list, target_pair_mapping_df, from_col=\"targets\", to_col=\"shorthand\" ):\n",
    "    cur_names = target_pair_mapping_df.loc[:,from_col].to_list()\n",
    "    new_names = target_pair_mapping_df.loc[:,to_col].to_list()\n",
    "    print()\n",
    "    result = target_pair_list\n",
    "    for i in range(len(cur_names)):\n",
    "        cur_name = cur_names[i]\n",
    "        new_name = new_names[i]\n",
    "        result = [target_pair.replace(cur_name, new_name) for target_pair in result]\n",
    "    return(result)\n",
    "def column_to_rownames(wgc, var=\"pos\"):\n",
    "    pos_list_full = wgc[var]\n",
    "    wgc = wgc.set_index(var)\n",
    "    return wgc\n",
    "def remove_model_in_param_grid(best_params):\n",
    "    results = {}\n",
    "    for param in best_params.keys():\n",
    "        new_param = param.replace(\"model__\", \"\")\n",
    "        results[new_param] = best_params[param]\n",
    "    return results\n",
    "def calculate_effect_size(x, y, method=\"slope\"):\n",
    "    # print(method)\n",
    "    if method == \"slope\":\n",
    "        result = (y[-1]-y[0])/(x[-1]-x[0])\n",
    "    elif method == \"auc\":\n",
    "        sorted_indices = np.argsort(x)\n",
    "        x_sorted = x[sorted_indices]\n",
    "        y_sorted = y[sorted_indices]\n",
    "\n",
    "        # Calculate area under the curve\n",
    "        result = np.trapz(y_sorted, x_sorted)\n",
    "    elif method == \"end\":\n",
    "        result = y[-1]\n",
    "    return result\n",
    "\n",
    "def get_average_lines(pdp_results, target_pair):\n",
    "    all_arrays = []\n",
    "    for random_seed in pdp_results.keys():\n",
    "        individual_lines = pdp_results[random_seed][\"coding_all\"][target_pair][\"individual\"]\n",
    "        all_arrays.append(individual_lines[0])\n",
    "    stacked = np.vstack(all_arrays)\n",
    "    return stacked"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "142521f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = {}\n",
    "model_designs = [\"rnaseq_vs_hiplex\"]\n",
    "frag_types = [\"mixed\"]\n",
    "cluster_result = pd.read_csv(\"/dcs05/hongkai/data/next_cutntag/bulk/explainability/leave_one_out/V/all/coding_cpg_clusters=5.csv\", index_col=0)\n",
    "cluster_result = cluster_result.set_index(\"gene_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a33b510e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load data\n",
    "data = {}\n",
    "model_designs = [\"rnaseq_vs_hiplex_rm_outlier_log\"]\n",
    "model_design = \"rnaseq_vs_hiplex_rm_outlier_log\"\n",
    "frag_types = [\"mixed\"]\n",
    "frag_type = \"mixed\"\n",
    "gene_select_names = [\"coding_all\"]\n",
    "gene_select_name = \"coding_all\"\n",
    "\n",
    "# read in rnaseq TPM\n",
    "rnaseq_dir = \"/dcs05/hongkai/data/next_cutntag/bulk/RNA-seq\"\n",
    "rnaseq_filename = \"RNA_seq_TPM_all.csv\"\n",
    "rnaseq_dir_filename = os.path.join(rnaseq_dir, rnaseq_filename)\n",
    "rnaseq = pd.read_csv(rnaseq_dir_filename, sep=\",\", header=0, index_col=0)\n",
    "# RNAseq is log transformed\n",
    "rnaseq.loc[:, \"sqrt_V\"] = np.log10(rnaseq.loc[:, \"V1V2\"]+1)\n",
    "\n",
    "# read in hiplex norm counts\n",
    "wgc_root_dir = \"/dcs05/hongkai/data/next_cutntag/bulk/wgc\"\n",
    "frag_type = \"mixed\"\n",
    "bin_size = \"promoter_-1000-1000\"\n",
    "scen = \"V\"\n",
    "target_qc_type = \"all\"\n",
    "post_process = \"libnorm\"\n",
    "wgc_dir = os.path.join(wgc_root_dir, frag_type, bin_size)\n",
    "wgc_filename = f\"{scen}_{frag_type}_{bin_size}_colQC-{target_qc_type}_{post_process}.feather\"\n",
    "wgc_dir_filename = os.path.join(wgc_dir, wgc_filename)\n",
    "wgc_raw = feather.read_feather(wgc_dir_filename)\n",
    "with open('/dcs05/hongkai/data/next_cutntag/script/utils/filtered_target_pairs.json', 'r') as file:\n",
    "    features = json.load(file)\n",
    "wgc_vals = wgc_raw[features]\n",
    "zero_hiplex_genes = wgc_raw[\"pos\"][wgc_vals.sum(axis=1) == 0].to_list()\n",
    "wgc_raw[features] = np.log10(wgc_raw[features]+1)\n",
    "wgc_raw[features] = (wgc_raw[features] - wgc_raw[features].min()) / (wgc_raw[features].max() - wgc_raw[features].min())\n",
    "\n",
    "features = [\"pos\"]+features\n",
    "wgc_raw = wgc_raw.loc[:, features]\n",
    "\n",
    "wgc_raw_list = wgc_raw.loc[:, \"pos\"].values.tolist()\n",
    "overlap_gene_list = list(set(wgc_raw_list) & set(gene_select_list))\n",
    "rnaseq_avail = rnaseq.loc[overlap_gene_list, [\"gene_id\", \"sqrt_V\"]]\n",
    "# filter genes with extremely high gene expression\n",
    "with open(\"/dcs05/hongkai/data/next_cutntag/bulk/explainability/rnaseq_hiplex_cutoff.json\", 'r') as f:\n",
    "    rnaseq_cutoffs = json.load(f)\n",
    "q99 = rnaseq_cutoffs[model_design]\n",
    "rnaseq_avail = rnaseq_avail.loc[rnaseq_avail.loc[:, \"sqrt_V\"] <= q99,:]\n",
    "zero_rnaseq_genes = rnaseq_avail[\"gene_id\"][rnaseq_avail[\"sqrt_V\"]==0].to_list()\n",
    "zero_all_genes = list(set(zero_hiplex_genes) & set(zero_rnaseq_genes))\n",
    "rnaseq_wgc_raw = rnaseq_avail.merge(wgc_raw, left_on=\"gene_id\", right_on=\"pos\", how=\"inner\")\n",
    "rnaseq_wgc_raw = rnaseq_wgc_raw.set_index(\"gene_id\")\n",
    "rnaseq_wgc_raw = rnaseq_wgc_raw.loc[~rnaseq_wgc_raw.index.isin(zero_all_genes), :]\n",
    "rnaseq_wgc = rnaseq_wgc_raw.drop(columns=[\"pos\"], inplace=False)\n",
    "rnaseq_wgc_train, rnaseq_wgc_test = train_test_split(rnaseq_wgc, test_size=0.2, random_state=42)\n",
    "rnaseq_wgc_train_X = rnaseq_wgc_train.drop(columns=[\"sqrt_V\"], inplace=False)\n",
    "rnaseq_wgc_train_y = rnaseq_wgc_train.loc[:, \"sqrt_V\"].values\n",
    "rnaseq_wgc_test_X = rnaseq_wgc_test.drop(columns=[\"sqrt_V\"], inplace=False)\n",
    "rnaseq_wgc_test_y = rnaseq_wgc_test.loc[:, \"sqrt_V\"].values\n",
    "rnaseq_wgc_all_X = rnaseq_wgc.drop(columns=[\"sqrt_V\"], inplace=False)\n",
    "rnaseq_wgc_all_y = rnaseq_wgc.loc[:, \"sqrt_V\"].values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3674e97",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# cross-validation\n",
    "model_design=\"rnaseq_vs_hiplex_rm_outlier_log\"\n",
    "gene_select_name =\"coding_all\"\n",
    "frag_type=\"mixed\"\n",
    "all_models={}\n",
    "all_models[\"coding_all\"] = {}\n",
    "model_params_file = f'/dcs05/hongkai/data/next_cutntag/script/explainability/{model_design}/{gene_select_name}.json'\n",
    "random_seed=42\n",
    "with open(model_params_file) as fp:\n",
    "    model_params = json.load(fp)\n",
    "    \n",
    "# set parameters\n",
    "for i in range(len(model_params)):\n",
    "    model_name = model_params[i]['model_name']\n",
    "    if model_name == \"rf\":\n",
    "        model = RandomForestRegressor(random_state=random_seed)\n",
    "    else:\n",
    "        model = LinearRegression()\n",
    "    params = model_params[i]['best params']\n",
    "\n",
    "    if len(params) > 0:\n",
    "        params = remove_model_in_param_grid(params)\n",
    "        print(params)\n",
    "        model.set_params(**params)\n",
    "    all_models[\"coding_all\"][model_name] = model\n",
    "\n",
    "print(gene_select_name)\n",
    "for frag_type in ['mixed']:\n",
    "    print(frag_type)\n",
    "    # rnaseq_wgc_all_X = data[gene_select_name][model_design][frag_type][\"rnaseq_wgc_all_X\"]\n",
    "    # rnaseq_wgc_all_y = data[gene_select_name][model_design][frag_type][\"rnaseq_wgc_all_y\"]\n",
    "    models = {\"random_forest\": all_models[gene_select_name][\"rf\"], \n",
    "              \"lr\": all_models[gene_select_name][\"lr\"]}\n",
    "    # 5-fold cross validation\n",
    "    kf = KFold(n_splits=5, shuffle=True, random_state=42)\n",
    "\n",
    "    cv_scores = {}\n",
    "    # Perform KFold cross-validation\n",
    "    for model_name, model in models.items():\n",
    "        rmse_scores = []\n",
    "        cors = []\n",
    "        spearman_cors = []\n",
    "        for train_index, test_index in kf.split(rnaseq_wgc_all_X):\n",
    "            X_train, X_test = rnaseq_wgc_all_X.iloc[train_index], rnaseq_wgc_all_X.iloc[test_index]\n",
    "            y_train, y_test = rnaseq_wgc_all_y[train_index], rnaseq_wgc_all_y[test_index]\n",
    "            \n",
    "            # Fit the model on training data\n",
    "            model.fit(X_train, y_train)\n",
    "            \n",
    "            # Predict on test data\n",
    "            y_pred = model.predict(X_test)\n",
    "            fig, ax = plt.subplots()\n",
    "            xy = np.vstack([y_pred,y_test])\n",
    "            z = gaussian_kde(xy)(xy)\n",
    "            print(model)\n",
    "            ax.scatter(y_pred, y_test, c=z, s=5)\n",
    "            # Create scatter plot\n",
    "            # plt.scatter(y_pred, y_test)\n",
    "            \n",
    "            # # Add labels and title\n",
    "            plt.xlabel('y_pred')\n",
    "            plt.ylabel('y_test')\n",
    "            plt.title(f'{model_name} {test_index}')\n",
    "            \n",
    "            # Show plot\n",
    "            plt.show()\n",
    "            # Calculate mean squared error\n",
    "            rmse = root_mean_squared_error(y_test, y_pred)\n",
    "            cor = scipy.stats.pearsonr(y_test, y_pred)[0]\n",
    "            spearman_cor = scipy.stats.spearmanr(y_test, y_pred)[0]\n",
    "            spearman_cors.append(spearman_cor)\n",
    "            rmse_scores.append(rmse)\n",
    "            cors.append(cor)\n",
    "        \n",
    "        average_spearman_cor = np.mean(spearman_cors)\n",
    "        average_rmse = np.mean(rmse_scores)\n",
    "        average_cor = np.mean(cors)\n",
    "        print(model_name)\n",
    "        print(f'Root Mean Squared Error for each fold: {rmse_scores}')\n",
    "        print(f'Average Mean Squared Error: {average_rmse}')\n",
    "        print(f'Pearson for each fold: {cors}')\n",
    "        print(f'Average Mean Pearson: {average_cor}\\nAverage Mean Spearman: {average_spearman_cor}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bce3eccb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# kmeans cluster\n",
    "model_design == \"rnaseq_vs_hiplex_rm_outlier_log\"\n",
    "wgc_raw = data[gene_select_name][model_design][frag_type][\"rnaseq_wgc_all_X\"]\n",
    "wcss = []\n",
    "for n_clusters in range(2, 11):\n",
    "    # kmeans = KMeans(n_clusters=i, init='k-means++', max_iter=300, n_init=10, random_state=0)\n",
    "    kmeans = KMeans(n_clusters=n_clusters, init='k-means++', random_state=42, max_iter=300, n_init=10)\n",
    "    kmeans.fit(wgc_raw)\n",
    "    wcss.append(kmeans.inertia_)  # inertia_ is the WCSS\n",
    "kn = \n",
    "(range(2, 11), wcss, curve='convex', direction='decreasing')\n",
    "print(kn.knee)\n",
    "best_n_clusters = kn.knee\n",
    "plt.plot(range(2, 11), wcss)\n",
    "plt.vlines(kn.knee, plt.ylim()[0], plt.ylim()[1], linestyles='dashed')\n",
    "plt.title('Elbow Method')\n",
    "plt.xlabel('Number of clusters')\n",
    "plt.ylabel('WCSS')\n",
    "plt.show()\n",
    "\n",
    "cluster_method = KMeans(n_clusters=best_n_clusters, random_state=42)\n",
    "# Fit the model\n",
    "cluster_method.fit(wgc_raw)\n",
    "# Assign cluster label\n",
    "wgc_raw.loc[:, \"cluster\"] = cluster_method.labels_\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec8db8af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# reorder cluster label based on average gene expression\n",
    "rnaseq_dir = \"/dcs05/hongkai/data/next_cutntag/bulk/RNA-seq\"\n",
    "rnaseq_filename = \"RNA_seq_TPM_all.csv\"\n",
    "rnaseq_dir_filename = os.path.join(rnaseq_dir, rnaseq_filename)\n",
    "rnaseq = pd.read_csv(rnaseq_dir_filename, sep=\",\", header=0, index_col=0)\n",
    "rnaseq.loc[:, \"sqrt_V\"] = rnaseq.loc[:, \"sqrt_V1V2\"]\n",
    "\n",
    "wgc_raw.loc[:, \"sqrt_V\"] = rnaseq.loc[wgc_raw.index, \"sqrt_V\"]\n",
    "gene_exp_order = wgc_raw.groupby('cluster')['sqrt_V'].mean().sort_values(ascending=True).index.to_list()\n",
    "swap_dict = dict(zip(\n",
    "    gene_exp_order,\n",
    "    list(range(len(gene_exp_order))),\n",
    "))\n",
    "print(wgc_raw.groupby('cluster')['sqrt_V'].mean().sort_values(ascending=True))\n",
    "print(swap_dict)\n",
    "wgc_raw['cluster'] = wgc_raw['cluster'].map(swap_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61621ec1",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.makedirs(f\"/dcs05/hongkai/data/next_cutntag/bulk/explainability/{model_design}/fig/coding_all/\", exist_ok=True)\n",
    "wgc_raw.to_csv(f\"/dcs05/hongkai/data/next_cutntag/bulk/explainability/{model_design}/fig/coding_all/cluster.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8da19008",
   "metadata": {},
   "outputs": [],
   "source": [
    "wgc_raw = pd.read_csv(f\"/dcs05/hongkai/data/next_cutntag/bulk/explainability/{model_design}/fig/coding_all/cluster.csv\", index_col=0)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d713c5a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_ids = wgc_raw['cluster'].unique()\n",
    "cluster_ids.sort()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0388c50",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_colors_df = pd.read_csv(\"/dcs05/hongkai/data/next_cutntag/bulk/explainability/rnaseq_vs_hiplex_rm_outlier_log/fig/feature_colors.csv\", index_col=0)\n",
    "feature_colors_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8dff9dc5",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_dir = f\"/dcs05/hongkai/data/next_cutntag/bulk/explainability/{model_design}/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28bc25e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from functools import reduce\n",
    "import pandas as pd\n",
    "\n",
    "# Assume df_list is your list of DataFrames\n",
    "# and they all have a common column called \"gene\"\n",
    "\n",
    "\n",
    "random_seeds = [0, 1, 7, 42, 123, 999, 1234, 1337, 2021, 31415]\n",
    "feature_importance_list = []\n",
    "for random_seed in random_seeds:\n",
    "    feature_importance = pd.read_csv(f\"{result_dir}/{random_seed}/top_features.csv\")\n",
    "    # short_target_pairs = map_target_names(feature_importance[\"feature\"], target_pair_mapping_df)\n",
    "    # feature_importance[\"feature\"] = short_target_pairs\n",
    "    feature_importance.rename(columns={'importance': f'{random_seed}_importance'}, inplace=True)\n",
    "    feature_importance_list.append(feature_importance)\n",
    "feature_importance_all = reduce(lambda left, right: pd.merge(left, right, on='feature'), feature_importance_list)\n",
    "feature_importance_all = feature_importance_all.set_index(\"feature\")\n",
    "feature_importance_all['importance'] = feature_importance_all.mean(axis=1)\n",
    "\n",
    "# Sort by row mean from biggest to smallest\n",
    "feature_importance = feature_importance_all.sort_values(by='importance', ascending=False)\n",
    "feature_importance = feature_importance.reset_index()\n",
    "feature_importance = feature_importance.loc[:19, [\"feature\", \"importance\"]]\n",
    "\n",
    "top_features = feature_importance[\"feature\"]\n",
    "short_target_pairs = map_target_names(feature_importance[\"feature\"], target_pair_mapping_df)\n",
    "feature_importance[\"feature\"] = short_target_pairs\n",
    "print(feature_importance)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9fa6400c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# feature importance plotting\n",
    "feature_importance[\"norm_feature_importance\"] = np.sqrt(feature_importance[\"importance\"])\n",
    "\n",
    "plt.figure(figsize=(6, 6))\n",
    "ax = sns.barplot(\n",
    "    data=feature_importance,\n",
    "    y='feature',\n",
    "    x='norm_feature_importance',\n",
    "    palette=feature_importance['colors'].tolist()\n",
    ")\n",
    "ax.grid(False)\n",
    "title_name = \"\"\n",
    "ax.set_title(title_name, fontsize=25, weight='bold')\n",
    "ax.set_xlabel('Feature Importance', fontsize=15)\n",
    "ax.set_ylabel('', fontsize=13)\n",
    "ax.tick_params(axis='x', labelsize=15)\n",
    "ax.tick_params(axis='y', labelsize=15)\n",
    "\n",
    "# Customize the x-axis tick labels to show original importance\n",
    "# Get current ticks, square them to invert the sqrt transform\n",
    "xticks = ax.get_xticks()\n",
    "xtick_labels = [f\"{x**2:.2f}\" for x in xticks]\n",
    "ax.set_xticklabels(xtick_labels)\n",
    "\n",
    "plt.tight_layout()\n",
    "os.makedirs(f\"/dcs05/hongkai/data/next_cutntag/bulk/explainability/{model_design}/fig/coding_all/\", exist_ok=True)\n",
    "plt.savefig(f\"/dcs05/hongkai/data/next_cutntag/bulk/explainability/{model_design}/fig/coding_all/feature_importance_scaled.pdf\", transparent=True)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e652160a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load pdp result calculated from 10 trained models\n",
    "result_dir = f\"/dcs05/hongkai/data/next_cutntag/bulk/explainability/{model_design}/\"\n",
    "\n",
    "# /dcs05/hongkai/data/next_cutntag/bulk/explainability/rnaseq_vs_hiplex/0/top_features_pdp.pkl\n",
    "random_seeds = [0, 1, 7, 42, 123, 999, 1234, 1337, 2021, 31415]\n",
    "pdp_results = {}\n",
    "for random_seed in random_seeds:\n",
    "    pdp_result_dir = f\"{result_dir}/{random_seed}/top_features_pdp.pkl\"\n",
    "    with open(pdp_result_dir, 'rb') as f:\n",
    "        pdp_lines_result = pickle.load(f)\n",
    "    pdp_results[random_seed] = pdp_lines_result\n",
    "random_seeds = [0, 1, 7, 42, 123, 999, 1234, 1337, 2021, 31415]\n",
    "cluster_results = []\n",
    "for random_seed in random_seeds:\n",
    "    rnaseq_wgc_all_X = pd.read_csv(f\"{result_dir}/{random_seed}/rnaseq_wgc_all_X.csv\", index_col=0)\n",
    "    rnaseq_wgc_all_X[\"cluster\"] = wgc_raw.loc[rnaseq_wgc_all_X.index, \"cluster\"]\n",
    "    rnaseq_wgc_all_X[\"category\"] = \"coding_cpg\"\n",
    "    rnaseq_wgc_all_X.loc[rnaseq_wgc_all_X.index.isin(gene_select_dict[\"coding_non_cpg\"]),\"category\"] = \"coding_non_cpg\"\n",
    "    cluster_results.append(rnaseq_wgc_all_X)\n",
    "cluster_results_stacked = pd.concat(cluster_results, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e73eb04f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# draw pdp plot for each epitope pair in each cluster\n",
    "effect_size_dict = {}\n",
    "gene_select_name = \"coding_all\"\n",
    "# Prepare the figure with subplots\n",
    "n_rows = len(top_features)  # one row per rm_target_pair\n",
    "n_cols = 2  # two columns: centered and not_centered\n",
    "fig, axes = plt.subplots(n_rows, n_cols, figsize=(7 * n_cols, 5 * n_rows), squeeze=False)\n",
    "# Iterate over plot types and features (rm_target_pair)\n",
    "for col_idx, plt_type in enumerate([\"centered\", \"not_centered\"]):\n",
    "    for row_idx, rm_target_pair in enumerate(top_features):\n",
    "        axis = axes[row_idx, col_idx]\n",
    "        print(f\"Plotting: gene_select_name={gene_select_name}, plt_type={plt_type}, rm_target_pair={rm_target_pair}\")\n",
    "        grid_values = pdp_results[random_seed][\"coding_all\"][rm_target_pair][\"grid_values\"][0]\n",
    "        x = grid_values\n",
    "        \n",
    "        axis.set_xlim(xmin=-0.1, xmax=1.5)\n",
    "        axis.set_xlabel('Hi-Plex Signal', fontsize=20)\n",
    "        axis.set_ylabel('Gene Expression', fontsize=20)\n",
    "\n",
    "        color_map = 'Paired'\n",
    "        cmap = plt.get_cmap(color_map)\n",
    "        colors = cmap(np.linspace(0, 1, len(cluster_ids)))\n",
    "        colors = [\"#D76532\", \"#F6C546\", \"#8574A4\", \"#5583C2\", \"#8CB463\"]\n",
    "        texts = []\n",
    "        for i, cluster_id in enumerate(cluster_ids):\n",
    "            line_color = colors[i]\n",
    "            individual_lines = get_average_lines(pdp_results, rm_target_pair)\n",
    "            individual_lines = individual_lines[cluster_results_stacked[\"cluster\"]==cluster_id,:]\n",
    "            y = individual_lines.mean(axis=0)\n",
    "            \n",
    "\n",
    "            if plt_type == \"centered\":\n",
    "                if min(grid_values) < 1:\n",
    "                    y_at_zero = y[np.argmin(np.abs(x))]\n",
    "                    y = y - y_at_zero\n",
    "                else:\n",
    "                    y = y  - y[0]\n",
    "            y = y[:len(x)]\n",
    "\n",
    "            axis.plot(\n",
    "                x,\n",
    "                y,\n",
    "                linewidth=1.8,\n",
    "                label=f'cluster {cluster_id}',\n",
    "                color=line_color\n",
    "            )\n",
    "            axis.tick_params(axis='x', labelsize=18)  # X-axis ticks\n",
    "            axis.tick_params(axis='y', labelsize=18)  # Y-axis ticks\n",
    "            texts.append(axis.text(x[-1] + 0.0125,\n",
    "                                   y[-1],\n",
    "                                   f'cluster {cluster_id+1}',\n",
    "                                   verticalalignment='center',\n",
    "                                   color=axis.get_lines()[-1].get_color(),\n",
    "                                   alpha=axis.get_lines()[-1].get_alpha(),\n",
    "                                   fontsize=22))\n",
    "        \n",
    "        adjust_text(texts, ax=axis)\n",
    "\n",
    "        axis.spines['top'].set_visible(False)\n",
    "        axis.spines['right'].set_visible(False)\n",
    "        rm_target_pair_short =map_target_names([rm_target_pair], target_pair_mapping_df)[0]\n",
    "        axis.set_title(f\"{rm_target_pair_short}\\n({plt_type})\", fontsize=28)\n",
    "        axis.grid(False)\n",
    "    # effect_size_dict[gene_select_name][plt_type] = effect_size_df\n",
    "\n",
    "plt.tight_layout()\n",
    "save_dir = f\"/dcs05/hongkai/data/next_cutntag/bulk/explainability/{model_design}/fig/pdp_grid/\"\n",
    "plt.show()\n",
    "os.makedirs(save_dir, exist_ok=True)\n",
    "fig.savefig(os.path.join(save_dir, f\"{gene_select_name}_pdp_grid.pdf\"))\n",
    "plt.close(fig)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28b6dc73",
   "metadata": {},
   "outputs": [],
   "source": [
    "# draw pdp plot for each epitope pair in each cluster with selected epitope pairs\n",
    "effect_size_dict = {}\n",
    "gene_select_name = \"coding_all\"\n",
    "# Prepare the figure with subplots\n",
    "n_rows = len(top_features)  # one row per rm_target_pair\n",
    "n_cols = 2  # two columns: centered and not_centered\n",
    "# fig, axes = plt.subplots(n_rows, n_cols, figsize=(7 * n_cols, 5 * n_rows), squeeze=False)\n",
    "target_pair_threshold = cluster_result.loc[:,~cluster_result.columns.isin([\"sqrt_V\", \"pos\", \"cluster\", \"category\"])].quantile(1)\n",
    "# Iterate over plot types and features (rm_target_pair)\n",
    "\n",
    "for col_idx, plt_type in enumerate([\"centered\"]):\n",
    "    color_map = 'Paired'\n",
    "    cmap = plt.get_cmap(color_map)\n",
    "    colors = cmap(np.linspace(0, 1, len(cluster_ids)))\n",
    "    for i, cluster_id in enumerate(cluster_ids):\n",
    "        fig, axis = plt.subplots(figsize=(6, 6))  # create a single subplot\n",
    "        axis.set_ylim(ymin=-0.5, ymax=0.75)\n",
    "        for gene_select_name in [\"coding_all\"]:\n",
    "            colors = [\"#D76532\", \"#F6C546\", \"#8574A4\", \"#4C9F70\"]\n",
    "            for row_idx, rm_target_pair in enumerate([\"H3K27me3-H3K4me3\", \"H3K4me1-H3K4me3\", \"H3K27ac-H3K4me3\", \"H3K27me3-H3K9me3\"]):\n",
    "                # pdp_lines_dict_i = pdp_lines_result[gene_select_name]\n",
    "                grid_values = pdp_results[random_seed][\"coding_all\"][\"H3K4me3-H3K4me3\"][\"grid_values\"][0]\n",
    "                grid_values = pdp_results[random_seed][\"coding_all\"][rm_target_pair][\"grid_values\"][0]\n",
    "                x = grid_values\n",
    "                x = x[x <= target_pair_threshold[rm_target_pair]]\n",
    "                if min(grid_values) < 0:\n",
    "                    axis.set_xlim(xmin=-4, xmax=4)\n",
    "                    axis.set_xlabel('log2FC(Hi-Plex Signal)', fontsize=20)\n",
    "                    axis.set_ylabel('log2FC(Gene Expression)', fontsize=20)\n",
    "                else:\n",
    "                    axis.set_xlim(xmin=-0.05, xmax=1.5)\n",
    "                    axis.set_xlabel('Hi-Plex Signal', fontsize=20)\n",
    "                    axis.set_ylabel('Gene Expression', fontsize=20)\n",
    "                texts = []\n",
    "                print(f\"Plotting: gene_select_name={gene_select_name}, plt_type={plt_type}, rm_target_pair={rm_target_pair}\")\n",
    "                line_color = colors[row_idx]\n",
    "                individual_lines = get_average_lines(pdp_results, rm_target_pair)\n",
    "                individual_lines = individual_lines[(cluster_results_stacked[\"cluster\"]==cluster_id),:]\n",
    "                y = individual_lines.mean(axis=0)\n",
    "                if plt_type == \"centered\":\n",
    "                    if min(grid_values) < 1:\n",
    "                        y_at_zero = y[np.argmin(np.abs(x))]\n",
    "                        y = y - y_at_zero\n",
    "                    else:\n",
    "                        y = y  - y[0]\n",
    "                y = y[:len(x)]\n",
    "                y = np.sign(y) * np.sqrt(np.abs(y))\n",
    "                linestyle = '-'\n",
    "                if gene_select_name == \"coding_non_cpg\":\n",
    "                    linestyle='--'\n",
    "                axis.plot(\n",
    "                    x,\n",
    "                    y,\n",
    "                    linewidth=1.8,\n",
    "                    label=f'cluster {cluster_id}',\n",
    "                    color=line_color,\n",
    "                    linestyle=linestyle\n",
    "                )\n",
    "                axis.tick_params(axis='x', labelsize=18)  # X-axis ticks\n",
    "                axis.tick_params(axis='y', labelsize=18)  # Y-axis ticks\n",
    "                if gene_select_name == \"coding_all\":\n",
    "                    texts.append(axis.text(x[-1] - 0.5,\n",
    "                                           y[-1]+ 0.025,\n",
    "                                           f'{rm_target_pair}',\n",
    "                                           verticalalignment='center',\n",
    "                                           color=axis.get_lines()[-1].get_color(),\n",
    "                                           alpha=axis.get_lines()[-1].get_alpha(),\n",
    "                                           fontsize=18))\n",
    "        \n",
    "        adjust_text(texts, ax=axis)\n",
    "        \n",
    "        def inv_signed_sqrt(y):\n",
    "            return np.sign(y) * (y ** 2)\n",
    "        from matplotlib.ticker import FuncFormatter\n",
    "\n",
    "        axis.yaxis.set_major_formatter(FuncFormatter(lambda val, pos: f\"{inv_signed_sqrt(val):.2f}\"))\n",
    "        axis.spines['top'].set_visible(False)\n",
    "        axis.spines['right'].set_visible(False)\n",
    "        rm_target_pair_short =map_target_names([rm_target_pair], target_pair_mapping_df)[0]\n",
    "        axis.set_title(f\"Cluster {cluster_id+1}\", fontsize=28)\n",
    "        axis.grid(False)\n",
    "        plt.tight_layout()\n",
    "        save_dir = f\"/dcs05/hongkai/data/next_cutntag/bulk/explainability/{model_design}/fig/coding_all/sqrt_coding_all_{plt_type}_same_range/\"\n",
    "        plt.show()\n",
    "        os.makedirs(save_dir, exist_ok=True)\n",
    "        fig.savefig(os.path.join(save_dir, f\"{cluster_id}.pdf\"))\n",
    "        plt.close(fig)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a0e2439",
   "metadata": {},
   "outputs": [],
   "source": [
    "# PDP plots for all epitope pairs\n",
    "width=10\n",
    "height=10\n",
    "model_design=\"rnaseq_vs_hiplex_rm_outlier_log\"\n",
    "ys_final = []\n",
    "for gene_select_name in [\"coding_all\"]:\n",
    "    for model_name in [\"random_forest\"]:\n",
    "        fig = plt.figure(figsize=(width,height))\n",
    "        axis1 = fig.add_subplot(111)\n",
    "        if model_design == \"rnaseq_vs_hiplex\":\n",
    "            axis1.set_xlim(xmin=-0.1, xmax=1.5)\n",
    "            axis1.set_ylim(ymin=-0.5, ymax=1.25)\n",
    "        else:\n",
    "            axis1.set_xlim(xmin=-4.5, xmax=4.5)\n",
    "        axis1.set_xlim(xmin=-0.1, xmax=1.5)\n",
    "        axis1.set_ylim(ymin=-0.5, ymax=1.25)\n",
    "        axis1.set_ylim(ymin=-0.1, ymax=0.1)\n",
    "        # model = all_models[model_design][gene_select_name][model_name]\n",
    "        target_pairs = list(top_features)\n",
    "        short_target_pairs = map_target_names(target_pairs, target_pair_mapping_df)\n",
    "        color_map = 'nipy_spectral'\n",
    "        cmap = plt.get_cmap(color_map)\n",
    "        cmap = cmap.reversed()\n",
    "        colors = cmap(np.linspace(0, 1, len(target_pairs)))\n",
    "        colors = [\n",
    "            \"#1b9e77\",  # dark teal green\n",
    "            \"#d95f02\",  # burnt orange\n",
    "            \"#7570b3\",  # muted indigo\n",
    "            \"#e7298a\",  # deep magenta\n",
    "            \"#66a61e\",  # olive green\n",
    "            \"#e6ab02\",  # golden mustard\n",
    "            \"#a6761d\",  # brown ochre\n",
    "            \"#666666\",  # dark gray\n",
    "            \"#1f78b4\",  # steel blue\n",
    "            \"#6a3d9a\",  # deep purple\n",
    "            \"#b15928\",  # rust brown\n",
    "            \"#01665e\",  # forest teal\n",
    "            \"#8c510a\",  # dark amber brown\n",
    "            \"#2c3e50\",  # dark slate blue-gray\n",
    "            \"#4b0082\",   # indigo (very dark purple)\n",
    "            \"#003f5c\",  # deep navy blue\n",
    "            \"#2f4f4f\",  # dark slate gray\n",
    "            \"#800000\",  # maroon\n",
    "            \"#191970\",  # midnight blue\n",
    "            \"#3c1053\",  # deep violet\n",
    "        ]\n",
    "        texts = []\n",
    "        effect_sizes = []\n",
    "        for i in range(len(target_pairs)):\n",
    "            target_pair = target_pairs[i]\n",
    "            target1 = target_pair.split(\"-\")[0]\n",
    "            target2 = target_pair.split(\"-\")[1]\n",
    "            linestyle = \"solid\"\n",
    "            # if target1 == target2:\n",
    "            #     linestyle = \"dashed\"\n",
    "            # else:\n",
    "            #     linestyle = \"solid\"\n",
    "            # pdp_lines_dict_i = pdp_lines_result[\"coding_all\"]\n",
    "            grid_values = pdp_results[random_seed][\"coding_all\"][target_pair][\"grid_values\"][0]\n",
    "            x = grid_values\n",
    "            individual_lines = get_average_lines(pdp_results, target_pair)\n",
    "            individual_lines = individual_lines\n",
    "            # print(individual_lines.shape)\n",
    "            y = individual_lines.mean(axis=0)\n",
    "            # print(y)\n",
    "            if model_design == \"rnaseq_vs_hiplex\":\n",
    "                y = y - y[0]\n",
    "            else:\n",
    "                y_at_zero = y[np.argmin(np.abs(x))]\n",
    "                y = y - y_at_zero\n",
    "            line_color = \"gray\"\n",
    "            ys_final.append(y[-1])\n",
    "            line_alpha = 0.8\n",
    "            \n",
    "            line_color = colors[i]\n",
    "            axis1.plot(\n",
    "                x,\n",
    "                y,\n",
    "                linewidth=2,\n",
    "                linestyle=linestyle,\n",
    "                label=target_pair,\n",
    "                color=line_color,\n",
    "                alpha=line_alpha\n",
    "            )\n",
    "            \n",
    "            texts.append(axis1.text(x[-1]+0.0125, \n",
    "                                            y[-1], \n",
    "                                            short_target_pairs[i], \n",
    "                                            verticalalignment='center', \n",
    "                                            color=axis1.get_lines()[-1].get_color(), \n",
    "                                            alpha=axis1.get_lines()[-1].get_alpha(),\n",
    "                                            fontsize=22))\n",
    "        adjust_text(texts, expand=(1.2, 2))\n",
    "        if model_design == \"rnaseq_vs_hiplex\":\n",
    "            axis1.set_xlabel('Hi-Plex Signal', fontsize=20)\n",
    "            axis1.set_ylabel('Gene Expression', fontsize=20)\n",
    "        else:\n",
    "            axis1.set_xlabel('log2(hiplex reads) FC', fontsize=20)\n",
    "            axis1.set_ylabel('log2(RNAseq TPM) FC', fontsize=20)\n",
    "        axis1.spines['top'].set_visible(False)\n",
    "        axis1.spines['right'].set_visible(False)\n",
    "        axis1.tick_params(axis='x', labelsize=18)  # X-axis ticks\n",
    "        axis1.tick_params(axis='y', labelsize=18)  # Y-axis ticks\n",
    "        # matplotx.line_labels()\n",
    "        # axis1.legend(target_pairs, loc='upper center', bbox_to_anchor=(0.5, -0.05), fancybox=True, shadow=True, ncol=5)\n",
    "        # axis1.title.set_text(f\"{gene_select_name} {model_name}\")\n",
    "        # axis1.set_title(f\"{model_design} {gene_select_name} {model_name}\", fontsize=25)\n",
    "        if gene_select_name == \"coding_all\":\n",
    "            title_name = \"Coding Genes\"\n",
    "        axis1.set_title(title_name, fontsize=28)\n",
    "        save_dir=f\"/dcs05/hongkai/data/next_cutntag/bulk/explainability/{model_design}/fig/{gene_select_name}/\"\n",
    "        os.makedirs(save_dir, exist_ok=True)\n",
    "        # axis1.set_title(f\"Partial Dependence Plot of Top-150 target Pairs \\non Gene Expression\", fontsize=28)\n",
    "        print(f'{save_dir}/summary_{width}x{height}.pdf')\n",
    "        # fig.savefig(f'{save_dir}/summary_{width}x{height}.pdf', bbox_inches = \"tight\", transparent=True)\n",
    "        # axis1.set_ylim([-0.1, 0.1])\n",
    "        # fig.savefig(f'{save_dir}/summary_zoom_in.pdf', bbox_inches = \"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1965d0e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# effect size calculation\n",
    "top_features = list(top_features)\n",
    "gene_select_name = 'coding_all'\n",
    "for effect_size_method in [\"auc\"]:\n",
    "# Iterate over plot types and features (rm_target_pair)\n",
    "    for col_idx, plt_type in enumerate([\"centered\"]):\n",
    "        effect_size_df = pd.DataFrame(np.zeros((len(cluster_ids), len(top_features))), index=cluster_ids, columns=top_features)\n",
    "        for row_idx, rm_target_pair in enumerate(top_features):\n",
    "            grid_values = grid_values = pdp_results[random_seed][\"coding_all\"][rm_target_pair][\"grid_values\"][0]\n",
    "            x = grid_values\n",
    "            for i, cluster_id in enumerate(cluster_ids):\n",
    "                line_color = colors[i]\n",
    "                individual_lines = get_average_lines(pdp_results, rm_target_pair)\n",
    "                if gene_select_name == \"coding_all\":\n",
    "                    individual_lines = individual_lines[cluster_results_stacked[\"cluster\"]==cluster_id,:]\n",
    "                else:\n",
    "                    individual_lines = individual_lines[(cluster_results_stacked[\"cluster\"]==cluster_id) & (cluster_results_stacked[\"category\"]==gene_select_name),:]\n",
    "                y = individual_lines.mean(axis=0)\n",
    "                if plt_type == \"centered\":\n",
    "                    if min(grid_values) < 1:\n",
    "                        y_at_zero = y[np.argmin(np.abs(x))]\n",
    "                        y = y - y_at_zero\n",
    "                    else:\n",
    "                        y = y  - y[0]\n",
    "                effect_size = calculate_effect_size(x, y, method=effect_size_method)\n",
    "                effect_size_df.loc[cluster_id, rm_target_pair] = effect_size\n",
    "        effect_size_df = effect_size_df / effect_size_df.std()\n",
    "        effect_size_df = effect_size_df.loc[effect_size_df.index.sort_values(),:]\n",
    "        short_target_pairs = map_target_names(effect_size_df.columns.tolist(), target_pair_mapping_df)\n",
    "        effect_size_df.columns = short_target_pairs\n",
    "        effect_size_df.index = effect_size_df.index + 1\n",
    "        effect_size_df.to_csv(f'{save_dir}/effect_size_{effect_size_method}.csv')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (next_cutntag)",
   "language": "python",
   "name": "next_cutntag"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
